homeassistant/components/sonos/switch.py

```diff
@@ -0,0 +1,216 @@
+            try:
+                _LOGGER.debug("Reached _discovered_player, soco=%s", soco)
+                for one_alarm in alarms.get_alarms(soco):
+                    if one_alarm.zone == soco and one_alarm not in alarm_list:
+                        _LOGGER.debug("Adding new alarm")
+                        alarm_list.append(one_alarm)
+                        hass.add_job(
+                            async_add_entities, [SonosAlarmSwitch(soco, one_alarm)],
+                        )
+            except SoCoException as ex:
+                _LOGGER.debug("SoCoException, ex=%s", ex)
+
+        if hosts:
+            for host in hosts:
+                try:
+                    _LOGGER.debug("Testing %s", host)
+                    player = pysonos.SoCo(socket.gethostbyname(host))
+                    if player.is_visible:
+                        # Make sure that the player is available
+                        _ = player.volume
+
+                        _discovered_alarm(player)
+                except (OSError, SoCoException) as ex:
+                    _LOGGER.debug("Exception %s", ex)
+                    if now is None:
+                        _LOGGER.warning("Failed to initialize '%s'", host)
+
+            _LOGGER.debug("Tested all hosts")
+            hass.helpers.event.call_later(DISCOVERY_INTERVAL, _discovery)
+        else:
+            _LOGGER.debug("Starting discovery thread")
+            pysonos.discover_thread(
+                _discovered_alarm,
+                interval=DISCOVERY_INTERVAL,
+                interface_addr=config.get(CONF_INTERFACE_ADDR),
+            )
+
+    _LOGGER.debug("Adding discovery job")
+    hass.async_add_executor_job(_discovery)
+
+
+class SonosAlarmSwitch(SwitchDevice):
+    """Switch class for Sonos alarms."""
+
+    def __init__(self, soco, alarm):
+        """Init Sonos alarms switch."""
+        self._icon = "mdi:alarm"
+        self._soco = soco
+        self._id = alarm._alarm_id
+        self._is_available = True
+        speaker_info = self._soco.get_speaker_info(True)
+        self._unique_id = "{}-{}".format(soco.uid, self._id)
+        self._name = "{} Alarm {}".format(speaker_info["zone_name"], self._id)
+        self._model = speaker_info["model_name"]
+
+        self.alarm = None
+        for one_alarm in alarms.get_alarms(self._soco):
+            # pylint: disable=protected-access
+            if one_alarm._alarm_id == self._id:
+                self.alarm = one_alarm
+
+        self._is_on = self.alarm.enabled
+        self._attributes = {
+            ATTR_TIME: str(self.alarm.start_time),
+            ATTR_VOLUME: str(self.alarm.volume),
+            ATTR_DURATION: str(self.alarm.duration),
+            ATTR_INCLUDE_LINKED_ZONES: str(self.alarm.include_linked_zones),
+            ATTR_RECURRENCE: str(self.alarm.recurrence),
+            ATTR_PLAY_MODE: str(self.alarm.play_mode),
+        }
+        super().__init__()
+
+    async def async_update(self, now=None):
```

Comment:
Soco is not async. Please use only `update`

Reply:
changed it